test/matrix_params.cc

```diff
@@ -32,3 +32,31 @@ void MatrixParams::mark()
     condD();
     seed();
 }
+
+//------------------------------------------------------------------------------
+/// Prints matrix kind and condition number.
+///
+/// @param[in] label
+///     Label to print describing matrix.
+///
+void MatrixParams::print( const char* label )
+{
+    if (kind.used()) {
+        printf( "%s: %s", label, kind().c_str() );
+
+        if (std::isnan( cond_used() ))
+            printf( ", cond unknown" );
+        else
+            printf( ", cond = %.5g", cond_used() );
+
+        // Warn if cond is requested (not NaN) but not used.
+        if (! std::isnan( cond() ) && cond() != cond_used())
+            printf( " (ignored requested cond = %.5g)", cond() );
+
+        // todo: condD used only by svd, heev, poev: warn if unused?
+        if (! std::isnan( condD() ))
+            printf( ", cond(D) = %.5g", condD() );
+
```

Comment:
Yes. Printing happens after calling the test routine, so we'll have to save the seed to print it.

Reply:
I ran into an issue when I printed the seed and then tried to use it to regenerate the same matrix.
```
Error on rank 0: --seed: invalid argument, 382834338537398656 outside [-1, 140737488355327]. (1 ranks had some error.)
```
In matrix_generator.cc, [configure_seed](https://github.com/icl-utk-edu/slate/blob/master/test/matrix_generator.cc#L1003-L1006) has:
```
1003        // higher order bits cannot affect the final seed
1004        user_seed %= int64_t(1) << 48;
1005        // scramble low order bits with a prime number less than 2^15
1006        user_seed *= 10477;
```
Line 1004 essentially chops it to 48 bits, right? Usually this is done with a bitmask, such as `user_seed &= 0xffffffffffff`. Compare:
```
In [45]: '%x %x' % (x % (1 << 48), x & 0xffffffffffff)
Out[45]: 'd6c86e8c7dfe d6c86e8c7dfe'
```
But then line 1006 makes it more than 48 bits again. Should the bitmask come 2nd, after line 1006? or perhaps be repeated?

The other issue is if the seed is printed, it prints the matrix for every single matrix size, which is very annoying:
```
pangolin slate/test> ./tester --matrix zero,identity,ij,orthog --dim 8,10,12 gemm
SLATE version 2022.07.00, id d2291f5b
input: ./tester --matrix zero,identity,ij,orthog --dim 8,10,12 gemm
2023-03-07 07:40:22, MPI size 1, OpenMP threads 4
                                                                                                                                                                                                                    
type  origin  target         gemm          go   transA   transB       m       n       k      alpha       beta    nb     p     q         la      error   time (s)       gflop/s  ref time (s)   ref gflop/s  status  
# matrix A: zero, cond = inf, seed = 408721798964170656
# matrix B: rand, cond unknown, seed = 408721800326180656
# matrix C: rand, cond unknown, seed = 408721800944323656
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   1.29e-16   1.40e-05        0.0731            NA            NA  pass    
# matrix A: zero, cond = inf, seed = 408721804736997656
# matrix B: rand, cond unknown, seed = 408721805082738656
# matrix C: rand, cond unknown, seed = 408721805449433656
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.28e-16   2.90e-05        0.0690            NA            NA  pass    
# matrix A: zero, cond = inf, seed = 408721808990659656
# matrix B: rand, cond unknown, seed = 408721809200199656
# matrix C: rand, cond unknown, seed = 408721809409739656
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.75e-16   3.00e-05         0.115            NA            NA  pass    
# matrix A: identity, cond = 1, seed = 408721811871834656
# matrix B: rand, cond unknown, seed = 408721812112805656
# matrix C: rand, cond unknown, seed = 408721812510931656
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   9.11e-17   2.80e-05        0.0366            NA            NA  pass    
# matrix A: identity, cond = 1, seed = 408721814983503656
# matrix B: rand, cond unknown, seed = 408721815402583656
# matrix C: rand, cond unknown, seed = 408721815716893656
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.21e-16   2.60e-05        0.0769            NA            NA  pass    
# matrix A: identity, cond = 1, seed = 408721818399005656
# matrix B: rand, cond unknown, seed = 408721818639976656
# matrix C: rand, cond unknown, seed = 408721818849516656
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   8.59e-17   3.00e-05         0.115            NA            NA  pass    
# matrix A: ij, cond unknown, seed = 408721821709737656
# matrix B: rand, cond unknown, seed = 408721822233587656
# matrix C: rand, cond unknown, seed = 408721822694575656
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   1.89e-16   4.30e-05        0.0238            NA            NA  pass    
# matrix A: ij, cond unknown, seed = 408721825628135656
# matrix B: rand, cond unknown, seed = 408721825911014656
# matrix C: rand, cond unknown, seed = 408721826078646656
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.24e-16   2.70e-05        0.0741            NA            NA  pass    
# matrix A: ij, cond unknown, seed = 408721828236908656
# matrix B: rand, cond unknown, seed = 408721828446448656
# matrix C: rand, cond unknown, seed = 408721828896959656
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.46e-16   2.10e-05         0.165            NA            NA  pass    
# matrix A: orthog, cond = 1, seed = 408721831254284656
# matrix B: rand, cond unknown, seed = 408721831788611656
# matrix C: rand, cond unknown, seed = 408721832113398656
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   1.73e-16   1.20e-05        0.0853            NA            NA  pass    
# matrix A: orthog, cond = 1, seed = 408721834826941656
# matrix B: rand, cond unknown, seed = 408721835246021656
# matrix C: rand, cond unknown, seed = 408721835539377656
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.49e-16   3.20e-05        0.0625            NA            NA  pass    
# matrix A: orthog, cond = 1, seed = 408721837865271656
# matrix B: rand, cond unknown, seed = 408721838095765656
# matrix C: rand, cond unknown, seed = 408721838263397656
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.70e-16   1.50e-05         0.230            NA            NA  pass    
All tests passed: gemm
```
versus currently:
```
pangolin slate/test> ./tester --matrix zero,identity,ij,orthog --dim 8,10,12 gemm
SLATE version 2022.07.00, id d2291f5b
input: ./tester --matrix zero,identity,ij,orthog --dim 8,10,12 gemm
2023-03-07 07:43:48, MPI size 1, OpenMP threads 4
                                                                                                                                                                                                                    
type  origin  target         gemm          go   transA   transB       m       n       k      alpha       beta    nb     p     q         la      error   time (s)       gflop/s  ref time (s)   ref gflop/s  status  
# matrix A: zero, cond = inf
# matrix B: rand, cond unknown
# matrix C: rand, cond unknown
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   8.36e-17   2.70e-05        0.0379            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.20e-16   2.00e-05         0.100            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.28e-16   3.10e-05         0.111            NA            NA  pass    
# matrix A: identity, cond = 1
# matrix B: rand, cond unknown
# matrix C: rand, cond unknown
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   9.93e-17   1.80e-05        0.0569            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.10e-16   3.80e-05        0.0526            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.33e-16   3.50e-05        0.0987            NA            NA  pass    
# matrix A: ij, cond unknown
# matrix B: rand, cond unknown
# matrix C: rand, cond unknown
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   9.39e-17   2.80e-05        0.0366            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.35e-16   1.70e-05         0.118            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.83e-16   2.50e-05         0.138            NA            NA  pass    
# matrix A: orthog, cond = 1
# matrix B: rand, cond unknown
# matrix C: rand, cond unknown
   d    host    task         auto         col  notrans  notrans       8       8       8   3.1+1.4i   2.7+1.7i   384     1     1          1   1.32e-16   3.30e-05        0.0310            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      10      10      10   3.1+1.4i   2.7+1.7i   384     1     1          1   1.96e-16   2.70e-05        0.0741            NA            NA  pass    
   d    host    task         auto         col  notrans  notrans      12      12      12   3.1+1.4i   2.7+1.7i   384     1     1          1   1.67e-16   4.00e-05        0.0864            NA            NA  pass    
All tests passed: gemm
```

Reply:
Yes, Line 1004 chops it to 48 bits. I was trying to avoid overflow in the following multiplication; although with PR #9 we could probably just remove the multiplication anyways.  (I used `%` instead of `&` since that's too many F's to easily count and the compiler should convert it to a bitmask anyways.)  

There is actually a second, implicit bit mask in [Lines 1014-1017](https://github.com/icl-utk-edu/slate/blob/master/test/matrix_generator.cc#L1014-L1017).  So, if we're printing the generated seed, we'd need to make that mask explicit.

Maybe when `seed()==-1`, we don't print the seed or print something like `seed unknown` or `seed = <generated>`.   That'd show the seeds when the user explicitly sets them, but it won't print when the user (presumably) doesn't care.  (And it'd avoid these issues with `configure_seed`.)

Reply:
So I'd like a mechanism that if a job runs and has a failure, I can exactly reproduce the line that failed. For instance, `gels` has sporadic failures, we think due to ill-conditioned matrices, but it's hard to reproduce. One thought is, `seed = -1` would set the initial seed based on the time stamp & print that out, but then all subsequent seeds are determined from that first one. Still unsure of the best way to extract this info.

Here's a mock-up.
```
pangolin slate/test> ./tester --dim 100 --method-gels cholqr --repeat 20 gels
% SLATE version 2022.07.00, id 1d7da13d
% input: ./tester --dim 100 --method-gels cholqr --repeat 20 gels
% 2023-05-15 11:39:21, MPI size 1, OpenMP threads 4
% seed = 1684166501124813000
                                                                                                                                                                                                             
type  origin  target  cholQR    gels   A   B   C    trans       m       n    nrhs    nb  ib    p    q  la  pt   leastsqr   min norm   residual   time (s)       gflop/s  ref time (s)   ref gflop/s  status  
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   8.32e-19         NA   7.13e-17   0.000712         2.327            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   3.89e-17         NA   2.51e-15   0.000730         2.270            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   8.97e-19         NA   8.85e-17   0.000732         2.263            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   3.71e-19         NA   4.27e-17   0.000663         2.499            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   2.35e-18         NA   2.85e-16   0.000551         3.007            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   2.37e-19         NA   3.41e-17   0.000593         2.794            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   8.38e-19         NA   7.95e-17   0.000486         3.409            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   4.60e-19         NA   4.64e-17   0.000496         3.340            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   1.23e-18         NA   1.55e-16   0.000357         4.641            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   1.46e-18         NA   1.54e-16   0.000335         4.946            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   4.06e-19         NA   4.59e-17   0.000380         4.360            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   7.52e-18         NA   3.78e-16   0.000349         4.747            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   1.52e-16         NA   1.01e-14   0.000418         3.964            NA            NA  FAILED  
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   5.11e-19         NA   5.97e-17   0.000330         5.021            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   8.57e-19         NA   1.45e-16   0.000363         4.564            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   8.18e-19         NA   9.04e-17   0.000358         4.628            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   1.20e-18         NA   9.72e-17   0.000354         4.680            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   2.52e-18         NA   1.91e-16   0.000364         4.552            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   1.92e-18         NA   1.13e-16   0.000347         4.775            NA            NA  pass    
   d    host    task    auto  cholqr   1   1   1  notrans     100     100      10   384  32    1    1   1   2   6.62e-19         NA   4.51e-17   0.000373         4.442            NA            NA  pass    


% Matrix kinds:
%  1: rand, cond unknown

% 1 tests FAILED: gels
```
Then somehow get the seed for that FAILED line to reproduce it later (I printed out seeds for every line, which really clutters up the output as previously discussed):
```
./tester --dim 100 --method-gels cholqr --repeat 20 --seed 1684166576865483000 --verbose 4 gels
...                                                                                                                                                                                                 
type  origin  target  cholQR    gels    trans       m       n    nrhs    nb  ib    p    q  la  pt   leastsqr   min norm   residual   time (s)       gflop/s  ref time (s)   ref gflop/s  status  
   d    host    task    auto  cholqr  notrans     100     100      10   384  32    1    1   1   2   1.52e-16         NA   1.01e-14   0.000667         2.484            NA            NA  FAILED  
```
I don't know the best way to go from printing a single seed at the top (`1684166501124813000`) to knowing the seed for a particular line (`1684166576865483000`).

Also easiest if there's just one seed that A, B, C (and other) matrices are all based on, rather than 3 separate seeds as it is now.

All of this can be handled in a separate PR.